File: python/tests/test_multi_field_sort.py
"""TC11 — Multi-field sort + sort combined with a filter.

The base filter suite only covers a single-field sort. This verifies a two-field
sort (primary key with ties broken by a secondary key) and sorting within a
filtered scope. Parity target for the other 8 suites.
"""

import pytest

from output import AND, Airtable, PrimaryModel
from output.static.table_helpers import SortOption

# Field-name sort keys (PrimaryField literals); use_field_ids defaults true on
# get(), but Airtable accepts field names in the sort param regardless.
NUMBER_INT = "Number (int)"
SINGLE_LINE_TEXT = "Single Line Text"


@pytest.fixture(scope="module")
def airtable():
    return Airtable()


def _row(number: int, text: str) -> PrimaryModel:
    m = PrimaryModel()
    m.primary_key = f"Sort {text}"
    m.number_int = number
    m.single_line_text = text
    return m


def _texts(records: list[PrimaryModel]) -> list[str]:
    return [r.single_line_text for r in records]


class TestMultiFieldSort:
    def test_two_field_sort_breaks_ties_on_secondary_key(self, airtable: Airtable):
        # number_int ties at 10 (rows "b" and "a"); secondary single_line_text orders them.
        created = airtable.primary.create([_row(10, "b"), _row(10, "a"), _row(20, "c")])
        ids = [r.id for r in created]
        try:
            sort: list[SortOption] = [
                {"field": NUMBER_INT, "direction": "asc"},
                {"field": SINGLE_LINE_TEXT, "direction": "asc"},
            ]
            results = airtable.primary.get(formula=PrimaryModel.f.id.in_list(ids), sort=sort)
            # (10,a), (10,b), (20,c) — tie on 10 broken by text asc.
            assert _texts(results) == ["a", "b", "c"]
        finally:
            airtable.primary.delete(record_ids=ids)

    def test_secondary_descending_reverses_tied_group(self, airtable: Airtable):
        created = airtable.primary.create([_row(10, "a"), _row(10, "b"), _row(20, "c")])
        ids = [r.id for r in created]
        try:
            sort: list[SortOption] = [
                {"field": NUMBER_INT, "direction": "asc"},
                {"field": SINGLE_LINE_TEXT, "direction": "desc"},
            ]
            results = airtable.primary.get(formula=PrimaryModel.f.id.in_list(ids), sort=sort)
            # number_int asc, then text DESC within the 10-tie: (10,b), (10,a), (20,c).
            assert _texts(results) == ["b", "a", "c"]
        finally:
            airtable.primary.delete(record_ids=ids)

    def test_sort_combined_with_a_filter(self, airtable: Airtable):
        created = airtable.primary.create([_row(30, "x"), _row(10, "y"), _row(20, "z"), _row(5, "low")])
        ids = [r.id for r in created]
        try:
            formula = AND(PrimaryModel.f.id.in_list(ids), PrimaryModel.f.number_int.gt(5))
            sort: list[SortOption] = [{"field": NUMBER_INT, "direction": "asc"}]
            results = airtable.primary.get(formula=formula, sort=sort)
            # Filtered to number_int > 5, sorted asc: 10(y), 20(z), 30(x). "low" filtered out.
            assert _texts(results) == ["y", "z", "x"]
        finally:
            airtable.primary.delete(record_ids=ids)

File: rust/tests/test_multi_field_sort.rs
use myairtable_tests::*;

// =============================================================================
// TC11 — Multi-field sort
//
// The base filter suite only covers a single-field sort. This verifies a
// two-field sort (primary key with ties broken by a secondary key) and sorting
// within a filtered scope. Mirrors csharp/tests/TestMultiFieldSort.cs.
// =============================================================================

fn setup() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let api_key = std::env::var("AIRTABLE_API_KEY").expect("AIRTABLE_API_KEY must be set");
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::new(&api_key, &base_id)
}

fn row(suite: &str, number: i64, text: &str) -> Fields {
    let mut f = Fields::new();
    f.set(PrimaryFields::PRIMARY_KEY_ID, format!("{suite} {text}"));
    f.set(PrimaryFields::NUMBER_INT_ID, number);
    f.set(PrimaryFields::SINGLE_LINE_TEXT_ID, text);
    f
}

fn scope_to(ids: &[&str]) -> String {
    if ids.len() == 1 {
        return format!("RECORD_ID()='{}'", ids[0]);
    }
    let parts: Vec<String> = ids.iter().map(|id| format!("RECORD_ID()='{id}'")).collect();
    format!("OR({})", parts.join(","))
}

fn texts(records: &[Record]) -> Vec<String> {
    records
        .iter()
        .map(|r| {
            r.fields
                .get(PrimaryFields::SINGLE_LINE_TEXT_ID)
                .unwrap()
                .as_str()
                .unwrap()
                .to_string()
        })
        .collect()
}

async fn try_delete_many(at: &Airtable, ids: &[String]) {
    if ids.is_empty() {
        return;
    }
    let _ = at.primary.dict.delete_many(ids).await;
}

// =============================================================================
// Two-field sort: ties on the primary key broken by the secondary key.
// =============================================================================

#[tokio::test]
async fn two_field_sort_breaks_ties_on_second_key() {
    let at = setup();
    let suite = "Sort TwoField";

    // NumberInt ties at 10 (rows "b" and "a"); the secondary SingleLineText sort orders them.
    let created = at
        .primary
        .dict
        .create_many(
            &[
                row(suite, 10, "b"),
                row(suite, 10, "a"),
                row(suite, 20, "c"),
            ],
            true,
        )
        .await
        .unwrap();
    let ids: Vec<String> = created.iter().map(|r| r.id.clone()).collect();
    let id_refs: Vec<&str> = ids.iter().map(|s| s.as_str()).collect();

    let params = AirtableQuery::new()
        .formula(scope_to(&id_refs))
        .sort(PrimaryFields::NUMBER_INT_ID, SortDirection::Asc)
        .sort(PrimaryFields::SINGLE_LINE_TEXT_ID, SortDirection::Asc);
    let results = at.primary.dict.get_many(&params).await.unwrap();

    // (10,a), (10,b), (20,c) — tie on 10 broken by text asc.
    assert_eq!(texts(&results), vec!["a", "b", "c"]);

    try_delete_many(&at, &ids).await;
}

// =============================================================================
// Secondary descending reverses the tied group.
// =============================================================================

#[tokio::test]
async fn secondary_descending_reverses_tied_group() {
    let at = setup();
    let suite = "Sort MixedDir";

    let created = at
        .primary
        .dict
        .create_many(
            &[
                row(suite, 10, "a"),
                row(suite, 10, "b"),
                row(suite, 20, "c"),
            ],
            true,
        )
        .await
        .unwrap();
    let ids: Vec<String> = created.iter().map(|r| r.id.clone()).collect();
    let id_refs: Vec<&str> = ids.iter().map(|s| s.as_str()).collect();

    let params = AirtableQuery::new()
        .formula(scope_to(&id_refs))
        .sort(PrimaryFields::NUMBER_INT_ID, SortDirection::Asc)
        .sort(PrimaryFields::SINGLE_LINE_TEXT_ID, SortDirection::Desc);
    let results = at.primary.dict.get_many(&params).await.unwrap();

    // NumberInt asc, then text DESC within the 10-tie: (10,b), (10,a), (20,c).
    assert_eq!(texts(&results), vec!["b", "a", "c"]);

    try_delete_many(&at, &ids).await;
}

// =============================================================================
// Sort combined with a filter.
// =============================================================================

#[tokio::test]
async fn sort_combined_with_a_filter() {
    let at = setup();
    let f = PrimaryModel::F;
    let suite = "Sort WithFilter";

    let created = at
        .primary
        .dict
        .create_many(
            &[
                row(suite, 30, "x"),
                row(suite, 10, "y"),
                row(suite, 20, "z"),
                row(suite, 5, "low"), // filtered out by NumberInt > 5
            ],
            true,
        )
        .await
        .unwrap();
    let ids: Vec<String> = created.iter().map(|r| r.id.clone()).collect();
    let id_refs: Vec<&str> = ids.iter().map(|s| s.as_str()).collect();

    let filter = formula::AND(&[&scope_to(&id_refs), &f.number_int.greater_than(5)]);
    let params = AirtableQuery::new()
        .formula(filter)
        .sort(PrimaryFields::NUMBER_INT_ID, SortDirection::Asc);
    let results = at.primary.dict.get_many(&params).await.unwrap();

    // Filtered to NumberInt > 5, sorted asc: 10(y), 20(z), 30(x).
    assert_eq!(texts(&results), vec!["y", "z", "x"]);

    try_delete_many(&at, &ids).await;
}
